pkg/kubelet/apis/pods/server.go

```diff
@@ -0,0 +1,508 @@
+	podsv1alpha1.UnimplementedPodsServer
+	lock        sync.RWMutex
+	pods        map[types.UID]*v1.Pod
+	broadcaster *broadcaster
+}
+
+// NewPodsServer creates a new PodServer for production use.
+func NewPodsServer(broadcaster *broadcaster) *PodsServer {
+	return &PodsServer{
+		pods:        make(map[types.UID]*v1.Pod),
+		broadcaster: broadcaster,
+	}
+}
+
+// NewPodsServerForTest creates a new PodServer with an injectable ticker for testing.
+func NewPodsServerForTest(broadcaster *broadcaster) *PodsServer {
+	return &PodsServer{
+		pods:        make(map[types.UID]*v1.Pod),
+		broadcaster: broadcaster,
+	}
+}
+
+// OnPodAdded is called when a pod is added.
+func (s *PodsServer) OnPodAdded(pod *v1.Pod) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	s.pods[pod.UID] = pod.DeepCopy()
+	s.broadcaster.Broadcast(PodWatchEvent{Type: watch.Added, Pod: pod})
+	logger := klog.FromContext(context.Background())
+	logger.Info("Pod added to storage", "podUID", pod.UID)
+}
+
+// OnPodUpdated is called when a pod is updated.
+func (s *PodsServer) OnPodUpdated(pod *v1.Pod) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	s.pods[pod.UID] = pod.DeepCopy()
+	s.broadcaster.Broadcast(PodWatchEvent{Type: watch.Modified, Pod: pod})
+	logger := klog.FromContext(context.Background())
+	logger.Info("Pod updated in storage", "podUID", pod.UID)
+}
+
+// OnPodRemoved is called when a pod is removed.
+func (s *PodsServer) OnPodRemoved(pod *v1.Pod) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	delete(s.pods, pod.UID)
+	s.broadcaster.Broadcast(PodWatchEvent{Type: watch.Deleted, Pod: pod.DeepCopy()})
+	logger := klog.FromContext(context.Background())
+	logger.Info("Pod removed from storage", "podUID", pod.UID)
+}
+
+// OnPodStatusUpdated is called when a pod's status is updated.
+func (s *PodsServer) OnPodStatusUpdated(pod *v1.Pod, status v1.PodStatus) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	if storedPod, ok := s.pods[pod.UID]; ok {
+		storedPod.Status = status
+		s.broadcaster.Broadcast(PodWatchEvent{Type: watch.Modified, Pod: storedPod})
+		logger := klog.FromContext(context.Background())
+		logger.Info("Pod status updated in storage", "podUID", pod.UID)
+	}
+}
+
+// Get returns a pod by UID.
+func (s *PodsServer) Get(uid types.UID) (*v1.Pod, bool) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+	pod, ok := s.pods[uid]
+	if !ok {
+		return nil, false
+	}
+	return pod.DeepCopy(), true
+}
+
+// List returns all pods.
+func (s *PodsServer) List() []*v1.Pod {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+	pods := make([]*v1.Pod, 0, len(s.pods))
+	for _, pod := range s.pods {
+		pods = append(pods, pod.DeepCopy())
+	}
+	sort.Slice(pods, func(i, j int) bool {
+		return pods[i].UID < pods[j].UID
+	})
+	return pods
+}
+
+// applyFieldMask uses Go's standard reflection to copy fields from src to dest based on the mask.
+// It supports traversing into slices to select specific fields from the elements.
+func ApplyFieldMask(mask *fieldmaskpb.FieldMask, src, dest interface{}) error {
```

Comment:
this looks expensive. Is any of it cache-able? Also individual field masks were not validated. And instead of invalid argument user will get simlpe erroras a return if they make a syntax error